From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tr7zw <tr7zw@live.de>
Date: Wed, 5 Aug 2020 08:05:10 -0500
Subject: [PATCH] Sugarcane configuration

also some basic settings that dont deserve a patch from Yatopia

diff --git a/src/main/java/co/aikar/timings/TimingsExport.java b/src/main/java/co/aikar/timings/TimingsExport.java
index 69d85e5a0a5ab9ffb682d8565aa08a4a567b68da..eb69dcc8d704b96879a72f633f4fdea7860bee78 100644
--- a/src/main/java/co/aikar/timings/TimingsExport.java
+++ b/src/main/java/co/aikar/timings/TimingsExport.java
@@ -232,8 +232,9 @@ public class TimingsExport extends Thread {
             pair("spigot", mapAsJSON(Bukkit.spigot().getSpigotConfig(), null)),
             pair("paper", mapAsJSON(Bukkit.spigot().getPaperConfig(), null)), // Tuinity - add config to timings report
             pair("tuinity", mapAsJSON(Bukkit.spigot().getTuinityConfig(), null)), // Tuinity - add config to timings report
-            pair("purpur", mapAsJSON(Bukkit.spigot().getPurpurConfig(), null))
+            pair("purpur", mapAsJSON(Bukkit.spigot().getPurpurConfig(), null)),
             // Purpur end
+            pair("sugarcane", mapAsJSON(Bukkit.spigot().getSugarcaneConfig(), null)) // Sugarcane - add config to timings report
         ));
 
         new TimingsExport(listeners, parent, history).start();
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 127e211548675155e8f3c11cf15775406c9f44bc..24e3865d6d7d712dd8484df225f69774bae08a38 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -235,7 +235,14 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         com.tuinity.tuinity.config.TuinityConfig.init((java.io.File) options.valueOf("tuinity-settings")); // Tuinity - Server Config
         gg.airplane.AirplaneConfig.load(); // Airplane - config
         gg.airplane.commands.AirplaneCommands.init(); // Airplane - command
-
+        // Sugarcane start
+        try {
+            org.sugarcanemc.sugarcane.config.SugarcaneConfig.init((java.io.File) options.valueOf("sugarcane-settings"));
+        } catch (Exception e) {
+            DedicatedServer.LOGGER.error("Unable to load server configuration", e);
+            return false;
+        }
+        // Sugarcane end
         this.setPvpAllowed(dedicatedserverproperties.pvp);
         this.setFlightAllowed(dedicatedserverproperties.allowFlight);
         this.setResourcePack(dedicatedserverproperties.resourcePack, this.getPackHash());
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 602b96d5dfbeb9a24072a8c80db3547a2af2ae6f..420a0011ef222169ea5b356d761164bd4debcc5a 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -269,7 +269,7 @@ public abstract class PlayerList {
         // Spigot - view distance
         playerconnection.send(new ClientboundLoginPacket(player.getId(), player.gameMode.getGameModeForPlayer(), player.gameMode.getPreviousGameModeForPlayer(), BiomeManager.obfuscateSeed(worldserver1.getSeed()), worlddata.isHardcore(), this.server.levelKeys(), this.registryHolder, worldserver1.dimensionType(), worldserver1.dimension(), this.getMaxPlayers(), worldserver1.getChunkSource().chunkMap.playerChunkManager.getLoadDistance(), flag1, !flag, worldserver1.isDebug(), worldserver1.isFlat())); // Paper - no-tick view distance // Tuinity - replace old player chunk management
         player.getBukkitEntity().sendSupportedChannels(); // CraftBukkit
-        playerconnection.send(new ClientboundCustomPayloadPacket(ClientboundCustomPayloadPacket.BRAND, (new FriendlyByteBuf(Unpooled.buffer())).writeUtf(this.getServer().getServerModName())));
+        playerconnection.send(new ClientboundCustomPayloadPacket(ClientboundCustomPayloadPacket.BRAND, (new FriendlyByteBuf(Unpooled.buffer())).writeUtf((org.sugarcanemc.sugarcane.config.SugarcaneConfig.brandName)))); // Sugarcane)));
         playerconnection.send(new ClientboundChangeDifficultyPacket(worlddata.getDifficulty(), worlddata.isDifficultyLocked()));
         playerconnection.send(new ClientboundPlayerAbilitiesPacket(player.getAbilities()));
         playerconnection.send(new ClientboundSetCarriedItemPacket(player.getInventory().selected));
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index d087131d8f5bc5fef10d41ca63daf4e2d9082c0d..a3250ec6f675bdd3b38096eb71f40f4b0f2de8cf 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -398,22 +398,24 @@ public abstract class LivingEntity extends Entity {
         this.level.getProfiler().push("livingEntityBaseTick");
         boolean flag = this instanceof net.minecraft.world.entity.player.Player;
 
-        if (this.isAlive()) {
-            if (this.isInWall()) {
-                this.hurt(DamageSource.IN_WALL, 1.0F);
-            } else if (flag && !this.level.getWorldBorder().isWithinBounds(this.getBoundingBox())) {
-                double d0 = this.level.getWorldBorder().getDistanceToBorder((Entity) this) + this.level.getWorldBorder().getDamageSafeZone();
-
-                if (d0 < 0.0D) {
-                    double d1 = this.level.getWorldBorder().getDamagePerBlock();
-
-                    if (d1 > 0.0D) {
-                        if (level.purpurConfig.teleportIfOutsideBorder && this instanceof ServerPlayer) { ((ServerPlayer) this).teleport(MCUtil.toLocation(level, level.getSharedSpawnPos())); return; } // Purpur
-                        this.hurt(DamageSource.IN_WALL, (float) Math.max(1, Mth.floor(-d0 * d1)));
+        if (!org.sugarcanemc.sugarcane.config.SugarcaneConfig.disableEntityStuckChecks) { // Sugarcane
+            if (this.isAlive()) {
+                if (this.isInWall()) {
+                    this.hurt(DamageSource.IN_WALL, 1.0F);
+                } else if (flag && !this.level.getWorldBorder().isWithinBounds(this.getBoundingBox())) {
+                    double d0 = this.level.getWorldBorder().getDistanceToBorder((Entity) this) + this.level.getWorldBorder().getDamageSafeZone();
+
+                    if (d0 < 0.0D) {
+                        double d1 = this.level.getWorldBorder().getDamagePerBlock();
+
+                        if (d1 > 0.0D) {
+                            if (level.purpurConfig.teleportIfOutsideBorder && this instanceof ServerPlayer) { ((ServerPlayer) this).teleport(MCUtil.toLocation(level, level.getSharedSpawnPos())); return; } // Purpur
+                            this.hurt(DamageSource.IN_WALL, (float) Math.max(1, Mth.floor(-d0 * d1)));
+                        }
                     }
                 }
             }
-        }
+        } // Sugarcane
 
         if (this.fireImmune() || this.level.isClientSide) {
             this.clearFire();
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 706dcbb278fb45722e8aecfd2e870943c4e51cbf..7a0885984846a2fcd13e715699b20605d88c23c3 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -871,6 +871,7 @@ public final class CraftServer implements Server {
             this.playerList.getBans().load();
         } catch (IOException ex) {
             this.logger.log(Level.WARNING, "Failed to load banned-players.json, " + ex.getMessage());
+            org.sugarcanemc.sugarcane.config.SugarcaneConfig.init((File) console.options.valueOf("sugarcane-settings")); // Sugarcane
         }
 
         org.spigotmc.SpigotConfig.init((File) console.options.valueOf("spigot-settings")); // Spigot
@@ -2410,6 +2411,13 @@ public final class CraftServer implements Server {
             org.spigotmc.RestartCommand.restart();
         }
 
+        // Sugarcane start
+        @Override
+        public YamlConfiguration getSugarcaneConfig() {
+            return org.sugarcanemc.sugarcane.config.SugarcaneConfig.config;
+        }
+        // Sugarcane end
+
         @Override
         public void broadcast(BaseComponent component) {
             for (Player player : CraftServer.this.getOnlinePlayers()) {
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 01ca38ee35f2c1e5031ea4b8aca09a2a59c4a475..32d661358b955fd37cb2a5313eb4f2c4aedc477d 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -162,6 +162,14 @@ public class Main {
                     .describedAs("Yml file");
                 // Purpur end
 
+                // Sugarcane start
+                acceptsAll(asList("sugarcane", "sugarcane-settings"), "File for Sugarcane settings")
+                        .withRequiredArg()
+                        .ofType(File.class)
+                        .defaultsTo(new File("sugarcane.yml"))
+                        .describedAs("Yml file");
+                // Sugarcane end
+
                 // Paper start
                 acceptsAll(asList("server-name"), "Name of the server")
                         .withRequiredArg()
diff --git a/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java b/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..5e2198ffad7d81ee25a717d0f111b332b3f9c95c
--- /dev/null
+++ b/src/main/java/org/sugarcanemc/sugarcane/config/SugarcaneConfig.java
@@ -0,0 +1,190 @@
+package org.sugarcanemc.sugarcane.config;
+
+import com.google.common.base.Throwables;
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.regex.Pattern;
+import org.bukkit.Bukkit;
+import org.bukkit.configuration.InvalidConfigurationException;
+import org.bukkit.configuration.file.YamlConfiguration;
+public class SugarcaneConfig {
+	public static File CONFIG_FILE;
+	private static final String HEADER = "This is the main configuration file for Sugarcane.\n"
+			+ "Sugarcane contains many breaking changes and settings, so know what you are doing!\n"
+			+ "You have been warned!\n"
+                       + "Join our Discord to receive support & optimization help: https://sugarcanemc.org/discord\n";
+	/*========================================================================*/
+	public static YamlConfiguration config;
+    public static int version; // since we're remapping sidestreams' configs we need this public
+    public static boolean verbose; // since we're remapping sidestreams' configs we need this public
+    /*========================================================================*/
+
+	public static void init(File configFile) {
+		CONFIG_FILE = configFile;
+		config = new YamlConfiguration();
+		try {
+			config.load(CONFIG_FILE);
+		} catch (IOException ex) {
+		} catch (InvalidConfigurationException ex) {
+			Bukkit.getLogger().log(Level.SEVERE, "Could not load sugarcane.yml, please correct your syntax errors", ex);
+			throw Throwables.propagate(ex);
+		}
+		config.options().header(HEADER);
+		config.options().copyDefaults(true);
+		verbose = getBoolean("verbose", false);
+		version = getInt("config-version", 1);
+		set("config-version", 1);
+		removeLeftovers();
+		readConfig(SugarcaneConfig.class, null);
+	}
+
+	private static void removeLeftovers() {
+		// this method is only to remove non-used values in the config
+
+		// leftover from rainforest
+		if (config.get("world-settings") != null) {
+			set("world-settings", null);
+		}
+		if (config.get("allow-player-item-duplication") != null) {
+			set("allow-player-item-duplication", null);
+		}
+		if (config.get("allow-ridable-chestable-duping") != null) {
+			set("allow-ridable-chestable-duping", null);
+		}
+		if (config.get("allow-sand-duping") != null) {
+			set("allow-sand-duping", null);
+		}
+	}
+
+	protected static void logError(String s) {
+		Bukkit.getLogger().severe(s);
+	}
+
+	protected static void log(String s) {
+		if (verbose) {
+			Bukkit.getLogger().info(s);
+		}
+	}
+
+	static void readConfig(Class<?> clazz, Object instance) {
+		for (Method method : clazz.getDeclaredMethods()) {
+			if (Modifier.isPrivate(method.getModifiers())) {
+				if (method.getParameterTypes().length == 0 && method.getReturnType() == Void.TYPE) {
+					try {
+						method.setAccessible(true);
+						method.invoke(instance);
+					} catch (InvocationTargetException ex) {
+						throw Throwables.propagate(ex.getCause());
+					} catch (Exception ex) {
+						Bukkit.getLogger().log(Level.SEVERE, "Error invoking " + method, ex);
+					}
+				}
+			}
+		}
+
+		try {
+			config.save(CONFIG_FILE);
+		} catch (IOException ex) {
+			Bukkit.getLogger().log(Level.SEVERE, "Could not save " + CONFIG_FILE, ex);
+		}
+	}
+
+	private static final Pattern SPACE = Pattern.compile(" ");
+		private static final Pattern NOT_NUMERIC = Pattern.compile("[^-\\d.]");
+
+	public static int getSeconds(String str) {
+	str = SPACE.matcher(str).replaceAll("");
+		final char unit = str.charAt(str.length() - 1);
+		str = NOT_NUMERIC.matcher(str).replaceAll("");
+		double num;
+		try {
+			num = Double.parseDouble(str);
+		} catch (Exception e) {
+			num = 0D;
+		}
+		switch (unit) {
+			case 'd':
+				num *= (double) 60 * 60 * 24;
+				break;
+				case 'h':
+				num *= (double) 60 * 60;
+				break;
+				case 'm':
+				num *= 60;
+				break;
+				default:
+				case 's':
+					break;
+		}
+		return (int) num;
+}
+
+	protected static String timeSummary(int seconds) {
+	String time = "";
+
+		if (seconds > 60 * 60 * 24) {
+			time += TimeUnit.SECONDS.toDays(seconds) + "d";
+			seconds %= 60 * 60 * 24;
+		}
+
+		if (seconds > 60 * 60) {
+			time += TimeUnit.SECONDS.toHours(seconds) + "h";
+			seconds %= 60 * 60;
+		}
+
+		if (seconds > 0) {
+			time += TimeUnit.SECONDS.toMinutes(seconds) + "m";
+		}
+		return time;
+}
+
+	private static void set(String path, Object val) {
+	config.set(path, val);
+}
+
+	private static boolean getBoolean(String path, boolean def) {
+	config.addDefault(path, def);
+		return config.getBoolean(path, config.getBoolean(path));
+}
+
+	private static double getDouble(String path, double def) {
+	 config.addDefault(path, def);
+		return config.getDouble(path, config.getDouble(path));
+}
+
+	private static float getFloat(String path, float def) {
+	 // TODO: Figure out why getFloat() always returns the default value.
+		return (float) getDouble(path, (double) def);
+}
+
+	private static int getInt(String path, int def) {
+	config.addDefault(path, def);
+		return config.getInt(path, config.getInt(path));
+}
+
+	private static <T> List<T> getList(String path, List<T> def) {
+	config.addDefault(path, def);
+		return (List<T>) config.getList(path, config.getList(path));
+}
+
+	private static String getString(String path, String def) {
+	config.addDefault(path, def);
+		return config.getString(path, config.getString(path));
+}
+
+	public static boolean disableEntityStuckChecks = false;
+private static void disableEntityStuckChecks() {
+	disableEntityStuckChecks = getBoolean("settings.disableEntityStuckChecks", false);
+}
+
+	public static String brandName = "Sugarcane";
+private static void brandName() {
+	brandName = getString("brand-name", brandName);
+    }
+}
\ No newline at end of file
